wpilibj/src/main/java/edu/wpi/first/wpilibj/experimental/controller/PIDController.java

```diff
@@ -0,0 +1,581 @@
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the Integral coefficient of the PID controller gain.
+   *
+   * @param Ki integral coefficient
+   */
+  @SuppressWarnings("ParameterName")
+  public void setI(double Ki) {
+    m_thisMutex.lock();
+    try {
+      m_Ki = Ki;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the Differential coefficient of the PID controller gain.
+   *
+   * @param Kd differential coefficient
+   */
+  @SuppressWarnings("ParameterName")
+  public void setD(double Kd) {
+    m_thisMutex.lock();
+    try {
+      m_Kd = Kd;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Get the Proportional coefficient.
+   *
+   * @return proportional coefficient
+   */
+  public double getP() {
+    m_thisMutex.lock();
+    try {
+      return m_Kp;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Get the Integral coefficient.
+   *
+   * @return integral coefficient
+   */
+  public double getI() {
+    m_thisMutex.lock();
+    try {
+      return m_Ki;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Get the Differential coefficient.
+   *
+   * @return differential coefficient
+   */
+  public double getD() {
+    m_thisMutex.lock();
+    try {
+      return m_Kd;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the period of this controller.
+   *
+   * @return the period of the controller.
+   */
+  public double getPeriod() {
+    return m_period;
+  }
+
+  /**
+   * Return the current controller output.
+   *
+   * <p>This is always centered around zero and constrained to the min and max outputs.
+   *
+   * @return The latest calculated output.
+   */
+  public double getOutput() {
+    m_thisMutex.lock();
+    try {
+      return m_output;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the reference for the PIDController.
+   *
+   * @param reference the desired reference
+   */
+  public void setReference(double reference) {
+    m_thisMutex.lock();
+    try {
+      if (m_maximumInput > m_minimumInput) {
+        m_reference = clamp(reference, m_minimumInput, m_maximumInput);
+      } else {
+        m_reference = reference;
+      }
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the current reference of the PIDController.
+   *
+   * @return the current reference
+   */
+  public double getReference() {
+    m_thisMutex.lock();
+    try {
+      return m_reference;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Return true if the error is within the percentage of the total input range,
+   * determined by SetTolerance. This asssumes that the maximum and minimum
+   * input were set using SetInput.
+   *
+   * <p>Currently this just reports on target as the actual value passes through the setpoint.
+   * Ideally it should be based on being within the tolerance for some period of time.
```

Comment:
It looks like this does use factor in period of time in tolerance.  Am I missing something?

Reply:
You're right. The velocity tolerance addressed this issue.